app-ktor-server/src/jvmMain/kotlin/com/ustadmobile/lib/rest/ResumableUploadRoute.kt

```diff
@@ -0,0 +1,104 @@
+package com.ustadmobile.lib.rest
+
+import com.ustadmobile.lib.rest.ResumableUploadRoute.SESSIONID
+import io.ktor.application.call
+import io.ktor.http.HttpHeaders
+import io.ktor.http.HttpStatusCode
+import io.ktor.request.ApplicationRequest
+import io.ktor.request.header
+import io.ktor.request.receiveStream
+import io.ktor.response.respond
+import io.ktor.response.respondText
+import io.ktor.routing.Route
+import io.ktor.routing.get
+import io.ktor.routing.put
+import io.ktor.routing.route
+import java.io.File
+import java.io.FileOutputStream
+import java.util.*
+
+
+/**
+ *  endpoint to create session
+ *  endpoint to send data
+ *  endpoint to return status
+ *
+ */
+fun Route.ResumableUploadRoute(folder: File) {
+
+    route("upload") {
+
+        get("createSession") {
+            val randomUuid = UUID.randomUUID()
+            File(folder, randomUuid.toString())
+            call.respondText(randomUuid.toString())
+        }
+
+        put("receiveData") {
+
+            val sessionId = call.request.header(SESSIONID) ?: ""
+
+            val (isValid, message) = call.request.isValidRequest(folder)
+            if (!isValid) {
+                call.respond(HttpStatusCode.BadRequest, message)
+            } else {
+
+                val input = call.receiveStream()
+                val uploadFile = File(folder, sessionId)
+                val fileOut = FileOutputStream(uploadFile, true)
```

Comment:
Fileoutputstream is not closed! I would suggest Kotlin's .use extension (which automatically closes something afterwards). It is also important to call flush to ensure everything was written. E.g.
```
FileOutputStream(uploadFile, true).use { fileOut ->
   fileOut.write(input.readBytes())
   fileOut.flush()
}
```